File: sdk/node/src/logs.ts
import type {
  NativeMachineLogChunk,
  NativeMachineLogHandle,
  NativeMachineLogOptionsInput,
  NativeMachineLogSource,
} from "./internal/napi.js";
import { mapNativePromise } from "./errors.js";
import type { MachineLogChunk } from "./types.js";
import { assertBoolean, assertRecord, assertUint8Array } from "./validation.js";

export function machineLogSourceToNative(source: unknown): NativeMachineLogSource {
  switch (source) {
    case "monitor":
    case "serial":
    case "exec":
    case "network":
    case "networkAudit":
      return source;
    default:
      throw new TypeError("source must be monitor, serial, exec, network, or networkAudit");
  }
}

export function machineLogOptionsToNative(
  options?: unknown,
): NativeMachineLogOptionsInput | undefined {
  if (options === undefined) return undefined;
  const record = assertRecord(options, "options");
  return {
    follow: record.follow === undefined ? undefined : assertBoolean(record.follow, "options.follow"),
  };
}

export function machineLogChunkFromNative(chunk: NativeMachineLogChunk): MachineLogChunk {
  const data = assertUint8Array(chunk.data, "machine log chunk.data");
  switch (chunk.output) {
    case "stdout":
    case "stderr":
      return { output: chunk.output, data };
    default:
      throw new TypeError("machine log chunk.output must be stdout or stderr");
  }
}

/** Async stream of semantic machine log chunks. */
export class MachineLogStream implements AsyncIterable<MachineLogChunk> {
  private closed = false;
  private receiving = false;

  constructor(private readonly native: NativeMachineLogHandle) {}

  /** Receive the next chunk, or `null` after the stream has ended. */
  async recv(): Promise<MachineLogChunk | null> {
    if (this.closed) throw new Error("machine log handle is closed");
    if (this.receiving) throw new Error("machine log handle is busy");

    this.receiving = true;
    try {
      const chunk = await mapNativePromise(this.native.recv());
      if (chunk === null) {
        this.closed = true;
        return null;
      }
      return machineLogChunkFromNative(chunk);
    } finally {
      this.receiving = false;
    }
  }

  /** Stop reading and release the native stream. Idempotent. */
  close(): void {
    if (this.closed) return;
    this.closed = true;
    this.native.close();
  }

  /** Iterate over raw machine log chunks until the stream ends. */
  async *[Symbol.asyncIterator](): AsyncIterator<MachineLogChunk> {
    try {
      while (true) {
        const chunk = await this.recv();
        if (chunk === null) return;
        yield chunk;
      }
    } finally {
      this.close();
    }
  }
}

File: sdk/go/logs.go
package silo

import (
	"context"
	"encoding/json"
	"github.com/vandycknick/silo/sdk/go/internal/ffi"
	"io"
	"sync"
)

type MachineLogSource string

const (
	MachineLogMonitor      MachineLogSource = "monitor"
	MachineLogSerial       MachineLogSource = "serial"
	MachineLogExec         MachineLogSource = "exec"
	MachineLogNetwork      MachineLogSource = "network"
	MachineLogNetworkAudit MachineLogSource = "network_audit"
)

type MachineLogOptions struct{ Follow bool }
type MachineLogOutput string

const (
	MachineLogStdout MachineLogOutput = "stdout"
	MachineLogStderr MachineLogOutput = "stderr"
)

type MachineLogChunk struct {
	Output MachineLogOutput `json:"output"`
	Data   []byte           `json:"data"`
}

func (m *Machine) Logs(ctx context.Context, source MachineLogSource, options MachineLogOptions) (*MachineLogStream, error) {
	if err := validateContext(ctx); err != nil {
		return nil, err
	}
	switch source {
	case MachineLogMonitor, MachineLogSerial, MachineLogExec, MachineLogNetwork, MachineLogNetworkAudit:
	default:
		return nil, newError(ErrorInvalidArgument, "", "unsupported machine log source")
	}
	request, _ := json.Marshal(struct {
		Source MachineLogSource `json:"source"`
		Follow bool             `json:"follow"`
	}{source, options.Follow})
	m.mutex.RLock()
	defer m.mutex.RUnlock()
	if m.closed {
		return nil, newError(ErrorClosed, "", "machine is closed")
	}
	native, err := m.native.Logs(request)
	if err != nil {
		return nil, fromNativeError(err)
	}
	return &MachineLogStream{native: native}, nil
}

// MachineLogStream reads one persisted semantic log source. Only one Recv may run at a time.
type MachineLogStream struct {
	mutex    sync.RWMutex
	receiver sync.Mutex
	native   *ffi.Log
	closed   bool
}

func (s *MachineLogStream) Recv(ctx context.Context) (*MachineLogChunk, error) {
	if err := validateContext(ctx); err != nil {
		return nil, err
	}
	if !s.receiver.TryLock() {
		return nil, newError(ErrorInvalidArgument, "", "machine log stream already has an active receiver")
	}
	defer s.receiver.Unlock()
	s.mutex.RLock()
	if s.closed {
		s.mutex.RUnlock()
		return nil, newError(ErrorClosed, "", "machine log stream is closed")
	}
	native := s.native
	s.mutex.RUnlock()
	done := make(chan struct {
		data *ffi.LogChunk
		eof  bool
		err  error
	}, 1)
	go func() {
		data, eof, err := native.Recv()
		done <- struct {
			data *ffi.LogChunk
			eof  bool
			err  error
		}{data, eof, err}
	}()
	select {
	case result := <-done:
		if result.err != nil {
			return nil, fromNativeError(result.err)
		}
		if result.eof {
			return nil, io.EOF
		}
		var output MachineLogOutput
		switch result.data.Output {
		case 1:
			output = MachineLogStdout
		case 2:
			output = MachineLogStderr
		default:
			return nil, newError(ErrorUnknown, "", "native log stream returned an unknown output channel")
		}
		return &MachineLogChunk{Output: output, Data: append([]byte(nil), result.data.Data...)}, nil
	case <-ctx.Done():
		_ = native.CloseStream()
		<-done
		return nil, contextError(ctx.Err())
	}
}
func (s *MachineLogStream) Close() error {
	if s == nil {
		return nil
	}
	s.mutex.RLock()
	if s.closed {
		s.mutex.RUnlock()
		return nil
	}
	native := s.native
	err := fromNativeError(native.CloseStream())
	s.mutex.RUnlock()

	s.receiver.Lock()
	defer s.receiver.Unlock()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if s.closed {
		return nil
	}
	native.Free()
	s.native = nil
	s.closed = true
	return err
}
